DOCSP-24551: POCOs page (#31)
mongodb/docs-csharp: source/fundamentals/poco.txt

.. _csharp-poco:

===============
Work with POCOs
===============

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

> In this guide, you can learn about how you can use "Plain Old CLR/Class

Author: should I include a wikipedia link?

Objects", or **POCOs**, with the {+driver-short+} for your operations
and queries. POCOs are simple class objects which do not inherit
features from any framework-specific base class and do not return any
special types. We recommend using POCOs in your {+language+} code to
adhere to idiomatic driver usage and achieve the best performance.

You should read this guide if you want to learn more about how to use
POCOs with the {+driver-short+} or need to adjust the driver's default
field mapping behavior.

Create a POCO
-------------

> You can create a POCO by instantiating any simple class that does not

Author: Agreed- I think I used instantiate correctly later in this doc but when first creating the class it should definitely be define.

implement interfaces or extend classes from a framework. When you
execute an operation such as a read or write using a POCO, the driver
interally *serializes*, or converts, the POCO to BSON.

Select the :guilabel:`POCO` or :guilabel:`BSON` tab to see how the
driver serializes a sample POCO to BSON:

.. tabs::

   .. tab:: POCO
      :tabid: poco-representation

      .. code-block:: csharp
         :copyable: false

         public class Clothing
         {
             public ObjectId _id { get; set; }
             public string name { get; set; }
             public bool instock { get; set; }
             public double price { get; set; }

>              public List<string> color { get; set; }

Author: If I use the C# naming conventions in the class, then when they are serialized to BSON, the fieldnames will also be capitalized/in PascalCase. Is that the tradeoff I should make? I'm not sure what you mean by adding a note.

>              public List<string> color { get; set; }

Author: edit-- I just kept PascalCase to follow C# and let the BSON also be in PascalCase. Instead i added the note about creating a `ConventionPack`. I think its better to follow the C# conventions, because any BSON conventions are not as strict.

>              public List<string> color { get; set; }

Reviewer: Sorry to be vague. I meant that you could keep the unconventional property names here (because, without `ConventionPack` or a Bson attribute, that's what's required for BSON serialization to work). We could then call out the fact that even though this example works, we can do better by correctly capitalizing the properties. 

In the **Set Field Names** section, what do you think about mentioning that BSON fields in the first example were PascalCase to match the C#, but should generally be camelCase--and that's the advantage of using BsonElement/ConventionPack?

         }

   .. tab:: BSON
      :tabid: bson-representation

      .. code-block:: json
         :copyable: false

         {
           "_id": ObjectId("..."),
           "name": "Long Sleeve Shirt",
           "instock": true,
           "price": 17.99,
           "color": [ "black", "navy", "red" ]
         }

You can define a POCO with any object structure that suits your
needs, including nested objects, arrays, lists, and any data types. The

> driver serializes POCOs using default mapping rules to generate the

Reviewer: My gut says that mentioning default mapping rules without explaining them will confuse some readers. When I reread that sentence, I wonder if we could just delete it altogether; I don't know that it adds much new information.

appropriate BSON representation of the document.

Custom Serialization
--------------------

If the default field mapping behavior does not meet your needs, you can
specify custom behavior using serialization related attributes. There
are various attributes which you can specify to modify the way that the
driver serializes each property of your POCO. This section describes
some of the common serialization related attributes.

> Set Field Names

Reviewer: Issue: `BsonElement` works for setting field names, but it onerous if there are lots of these fields.

Suggestion: In a tip box, mention that if all you want to do is map PascalCase properties to camelCase fields, you can use code like this in the program instead of `BsonElement`:

```
// This allows automapping of the camelCase database fields to our models. 
var camelCaseConvention = new ConventionPack { new CamelCaseElementNameConvention() };
ConventionRegistry.Register("CamelCase", camelCaseConvention, type => true);
```

~~~~~~~~~~~~~~~

The driver serializes POCO properties to BSON fields with same field
name and capitalization. To store a property under a different name, you

> can use the ``BsonElementAttribute``. The following code shows how you

Author: I will change this for all of the attributes, and in the API documentation section for consistency.

can direct the driver to store the ``YearBuilt`` property of the
``House`` object as ``year_built`` in the serialized BSON document:

.. code-block:: csharp
   :copyable: true

   public class House
   {
       public ObjectId Id { get; set; }

       [BsonElement("year_built")]
       public int YearBuilt { get; set; }
   }

Select Type Representation
~~~~~~~~~~~~~~~~~~~~~~~~~~

> For some {+language+} primitive types you can control the BSON type used

Author: it is unclear from the api docs, I will reword.

to represent the value of your POCO property using the

> ``BsonRepresentationAttribute``. The following code shows how you can
> specify that the ``YearBuilt`` property, defined as a ``char`` in
> {+language+}, is serialized as a BSON ``Int32`` type:

Reviewer: ```suggestion
``BsonRepresentationAttribute``. In the following code sample, the ``YearBuilt`` property, defined as a ``char`` in
{+language+}, is serialized as a BSON ``Int32`` type:
```

.. code-block:: csharp
   :copyable: true

   public class House
   {
       public ObjectId Id { get; set; }

       [BsonRepresentation(BsonType.Int32)]
       public char YearBuilt { get; set; }
   }

Set Field Order
~~~~~~~~~~~~~~~

The driver serializes properties to BSON fields in the order they

> are specified in the POCO. To store properties in a custom order, you

Reviewer: Issue: When would it be useful to use a custom order?

Suggestion: Link or explanation

Author: added some explanation

> can specify the ``Order`` named parameter in the ``BsonElement``

Reviewer: ```suggestion
can specify the ``Order`` named parameter in the ``[BsonElement()]``
```

> attribute. The following code shows how you can direct the driver to
> store the ``YearBuilt`` property after the ``Style`` property:

Reviewer: ```suggestion
attribute. In the following code sample, the driver stores the ``YearBuilt`` property after the ``Style`` property:
```

.. code-block:: csharp
   :copyable: true

   public class House
   {
       public ObjectId Id { get; set; }

       [BsonElement(Order = 2)]
       public int YearBuilt { get; set; }

       [BsonElement(Order = 1)]
       public string Style { get; set; }
   }

> Any properties that do not have an explicit ``Order`` will occur after
> those that do.

Reviewer: ```suggestion
If any properties don't have an explicit ``Order``, the driver will serialize them in the default order after
those that do.
```

> Identify Id Property

Reviewer: ```suggestion
Identify ``Id`` Property
```

~~~~~~~~~~~~~~~~~~~~

> By default, the driver uses any public property named ``Id``, ``id``, or ``_id`` as the

Reviewer: Issue: Do we know what happens if I have more than one such property?

Suggestion: Link or explanation, possibly a warning box if it's something we want users to avoid.

> By default, the driver uses any public property named ``Id``, ``id``, or ``_id`` as the
> identifier to map to the BSON ``_id`` field. You can explicitly select the

Reviewer: ```suggestion
By default, the driver maps any public property named ``Id``, ``id``, or ``_id`` to the BSON ``_id`` field. You can explicitly select the
```

> identifier to map to the BSON ``_id`` field. You can explicitly select the
> identifier using the ``BsonIdAttribute``. The following code shows how

Reviewer: ```suggestion
identifier to map to the BSON ``_id`` field. To explicitly select the
property to map to the ``_id`` field, use the ``[BsonId()]`` attribute. The following code shows how
```

> identifier using the ``BsonIdAttribute``. The following code shows how
> you can direct the driver to map the ``Identifier`` property to the ``_id`` field:

Reviewer: ```suggestion
identifier using the ``BsonIdAttribute``. The following code sample maps the ``Identifier`` property to the ``_id`` field:
```

.. code-block:: csharp
   :copyable: true

   public class House
   {
       [BsonId]
       public string Identifier { get; set; }
   }

Omit Empty Fields
~~~~~~~~~~~~~~~~~

> By default, the driver serializes undefined properties to fields with ``null``
> values. You can direct the driver to ignore empty properties using the

Reviewer: Issue: We call these properties "undefined" in the first sentence, then "empty" in the next. 

```suggestion
By default, the driver serializes undefined properties to fields with ``null``
values. To ignore undefined properties, use the
```

> ``BsonIgnoreAttribute``. The following code shows how you can prevent

Reviewer: ```suggestion
``[BsonIgnore()]`` attribute. The following code shows how you can prevent
```

the driver from serializing the ``YearBuilt`` property if it is empty:

.. code-block:: csharp
   :copyable: true

   public class House
   {
       public ObjectId Id { get; set; }

       [BsonIgnore]
       public int YearBuilt { get; set; }
       public string Style { get; set; }
   }

Example
-------

> The following example shows how the {+driver-short+} inserts a document
> instantiated as a POCO with the following custom field mapping specifications:

Reviewer: Issue: This feels a little long and confusing, but I also struggled with cleaning it up! 

```suggestion
The following example shows how to define a C# class with custom field mapping specifications, create a POCO to represent a document, then insert the POCO into MongoDB. 
```

> - Sets a custom field name ``in_stock`` for the ``instock`` property
> - Serializes the ``price`` field as a BSON ``Double`` type
> 
> Define the ``Clothing`` class with the relevant serialization
> attributes:

Reviewer: ```suggestion
The ``Clothing`` class uses serialization attributes to perform the following:
- Sets a custom field name ``in_stock`` for the ``instock`` property
- Serializes the ``price`` field as a BSON ``Double`` type
```

.. literalinclude:: ../includes/fundamentals/code-examples/poco.cs
   :start-after: start-model
   :end-before: end-model
   :language:  csharp
   :copyable:
   :dedent:

Instantiate a ``Clothing`` object and insert the document into a collection:

.. literalinclude:: ../includes/fundamentals/code-examples/poco.cs
   :start-after: start-insert
   :end-before: end-insert
   :language:  csharp
   :copyable:
   :dedent:

> The BSON representation of the inserted document should resemble the

Reviewer: Issue: Per the style guide, we try to avoid "should", as it suggests we're doubtful about what our code does :)

```suggestion
The BSON representation of the inserted document looks like this:
```

following:

.. code-block:: json
   :copyable: false

   {
     "_id": ObjectId("..."),
     "name": "Denim Jacket",
     "in_stock": false,
     "price": 32.99,
     "color": [ "dark wash", "light wash" ]
   }

Additional Information
----------------------

> For a full list of serialization related attributes, see the

Reviewer: ```suggestion
For a full list of serialization-related attributes, see the
```

`Serialization.Attributes API documentation <{+api-root+}/N_MongoDB_Bson_Serialization_Attributes.htm>`__.

For additional read and write operation examples using POCOs, see the :ref:`Usage Examples
<csharp-usage-examples>` or the :ref:`CRUD Fundamentals Pages <csharp-crud>`.

API Documentation
~~~~~~~~~~~~~~~~~

To learn more about any of the methods or types discussed in this

> guide, see the following API Documentation:

Reviewer: ```suggestion
guide, see the following API documentation:
```

the ``[BsonElement()]`` attribute. The following code maps the
``YearBuilt`` property of the ``House`` class to the ``year_built``
field in the serialized BSON document:

.. code-block:: csharp
   :copyable: true

   public class House
   {
       public ObjectId Id { get; set; }

       [BsonElement("year_built")]
       public int YearBuilt { get; set; }
   }

.. tip:: Set Custom Field Name Convention
   
   If you want to serialize every property with a custom field name, you
   can define a ``ConventionPack`` instead of using the
   ``[BsonElement()]`` attribute. For example, if your define your POCO
   using the PascalCase naming convention, you can use the following
   code to use camelCase field names in the serialized document:

   .. literalinclude:: ../includes/fundamentals/code-examples/poco.cs
      :start-after: start-conventionpack
      :end-before: end-conventionpack
      :language:  csharp
      :copyable:
      :dedent:

Select Type Representation
~~~~~~~~~~~~~~~~~~~~~~~~~~

You can control the BSON type that the driver serializes a property to using the
``[BsonRepresentation()]`` attribute if the {+language+} primitive type

> is convertible to the BSON type you specify. In the following code sample, the

Author: I changed the wording a little bit but broke the sentence up!

> is convertible to the BSON type you specify. In the following code sample, the

Author: I am including a link to the Microsoft C# programming guide.

match an existing schema, you can specify the ``Order`` named
parameter in the ``[BsonElement()]`` attribute. In the following code
sample, the driver stores the ``YearBuilt`` property after the
``Style`` property:

.. code-block:: csharp
   :copyable: true

   public class House
   {
       public ObjectId Id { get; set; }

       [BsonElement(Order = 2)]
       public int YearBuilt { get; set; }

       [BsonElement(Order = 1)]
       public string Style { get; set; }
   }

If any properties don't have an explicit ``Order``, the driver will
serialize them in the default order after those that do.

Identify ``Id`` Property
~~~~~~~~~~~~~~~~~~~~~~~~

By default, the driver maps any public property named ``Id``, ``id``, or
``_id`` to the BSON ``_id`` field. To explicitly select the
property to map to the ``_id`` field, use the ``[BsonId()]`` attribute.
The following code sample maps the ``Identifier`` property to the
``_id`` field:

.. code-block:: csharp
   :copyable: true

   public class House
   {
       [BsonId]
       public string Identifier { get; set; }
   }

.. warning:: Multiple Id Fields

   If you identify more than one property as the ``_id`` field, the

>    driver throws a ``BsonSerializationException``.

Author: updated the warning to address both cases.

Omit Empty Fields
~~~~~~~~~~~~~~~~~

By default, the driver serializes undefined properties to fields with ``null``
values. To ignore undefined properties during serialization, use the ``[BsonIgnore()]``
attribute. The following code shows how you can prevent the driver from
serializing the ``YearBuilt`` property if it is undefined:

.. code-block:: csharp
   :copyable: true

   public class House
   {
       public ObjectId Id { get; set; }

       [BsonIgnore]
       public int YearBuilt { get; set; }
       public string Style { get; set; }
   }

Example
-------

The following example shows how to define a {+language+} class with custom field
mapping specifications, create a POCO to represent a document, then
insert the document into MongoDB. The ``Clothing`` class uses
serialization attributes to perform the following:

- Sets a custom field name ``StyleName`` for the ``Name`` property
- Serializes the ``Price`` field as a BSON ``Double`` type

Define the ``Clothing`` class with the relevant serialization-related

> attributes:

Author: because the descriptions for what the serialization attributes do are kind of wordy, I like them in the list format. I reworded and shortened the section.

> attributes:

Reviewer: Looks fine except [we try to avoid introducing lists with sentence fragments.](https://www.mongodb.com/docs/meta/style-guide/style/lists/). That part should be reworded but I won't request changes on it. Let me know if you want to brainstorm together.

.. literalinclude:: ../includes/fundamentals/code-examples/poco.cs
   :start-after: start-model
   :end-before: end-model
   :language:  csharp
   :copyable:
   :dedent:

> Instantiate a ``Clothing`` object and insert the document into a collection:

Reviewer: ```suggestion
The `Program` class instantiates a ``Clothing`` object and inserts the document into a collection:
```

Author: reworded this and the preceding code intro to say "The following code..."
